steward/src/cork.rs

```diff
@@ -0,0 +1,144 @@
+use crate::{
+    cellars::{self, uniswapv3, CellarId},
+    error::{Error, ErrorKind},
+    prelude::APP,
+    somm_send,
+};
+use abscissa_core::{
+    tracing::log::{debug, error, warn},
+    Application,
+};
+use deep_space::{Coin, Contact};
+use gravity_bridge::gravity_proto::cosmos_sdk_proto::cosmos::base::abci::v1beta1::TxResponse;
+use somm_proto::cork::{query_client::QueryClient as CorkQueryClient, Cork, QueryCellarIDsRequest};
+use std::time::Duration;
+use steward_proto::{
+    self,
+    steward::{
+        self,
+        submit_request::ContractCallData::{self, Uniswapv3Rebalance},
+        SubmitRequest, SubmitResponse,
+    },
+};
+use tonic::{self, async_trait, Code, Request, Response, Status};
+
+pub struct CorkHandler;
+
+#[async_trait]
+impl steward::contract_call_server::ContractCall for CorkHandler {
+    async fn submit(
+        &self,
+        request: Request<SubmitRequest>,
+    ) -> Result<Response<SubmitResponse>, Status> {
+        debug!("received contract call request");
+        let request = request.get_ref();
+
+        // Check if cellar is governance approved before building cork
+        let config = APP.config();
+        let mut client = match CorkQueryClient::connect(config.cosmos.grpc.clone()).await {
+            Ok(c) => c,
+            Err(err) => {
+                error!("cork query client connection failed: {}", err);
+                return Err(Status::new(
+                    Code::Internal,
+                    "failed to query chain to validate cellar id",
+                ));
+            }
+        };
+        debug!("checking if cellar ID is approved");
+        let ids = &client
+            .query_cellar_i_ds(QueryCellarIDsRequest {})
```

Comment:
I think ideally we have a background thread running to update this state so we don't have to make this call every request.